libexec/revng/revng-check-conventions

```diff
@@ -1,498 +1,727 @@
-      projects where Qt coding conventions are a better fit).
+    @staticmethod
+    def combine(*pass_results: "PassResult") -> "PassResult":
+        return PassResult(
+            "\n".join(p.output for p in pass_results), all(p.ok for p in pass_results)
+        )
 
-      See clang-format documentation for more details.
 
-    --print-clang-format-config
-      Print the clang-format configuration to stdout, ignoring all the other
-      arguments except for --use-local-clang-format-file.
-      The conventions are not checked.
+class Pass(abc.ABC):
+    def __init__(self, name: str, tag: str):
+        self.name = name
+        self.tag = tag
 
-    --commit-range <RANGE>
-      Check files changed in the specified commit range
+    def __call__(self, files: Dict[str, Dict[Path, str]], args: Dict[str, str]) -> PassResult:
+        file_list = files[self.tag]
+        if len(file_list.keys()) == 0:
+            return PassResult("", True)
+        return self.process(files[self.tag], args)
 
-    --HEAD
-      Alias for '--commit-range HEAD^..HEAD'
+    @abc.abstractmethod
+    def process(self, files: Dict[Path, str], args: Dict[str, str]) -> PassResult:
+        ...
 
-    --cached
-      Only check staged files
 
-    --all
-      Check all files instead of only changed ones
+def template_iterable(iterable: Iterable[str], args: Dict[str, str]) -> List[str]:
+    return [Template(element).substitute(args) for element in iterable]
 
-    FILES
-      List of filenames of the files for which you want to check the rev.ng
-      coding conventions.
 
-RETURN VALUES
-   On success exit code is 0.
-   On failure, i.e. if there is at least one file that is not respecting the
-   coding conventions, exit code is 1.
-END_USAGE_TEXT
-}
+def run_with_capture(cmd: Sequence[str]) -> ProcessResult:
+    result = process_run(cmd, stdout=PIPE, stderr=STDOUT, text=True)
+    return ProcessResult(result.stdout, result.returncode)
 
-# git diff wrapper
-# Wraps git diff to extract only Added, Modified, Renamed and Copied files
-function git_diff_wrapper() {
-    local FIELDS RESULT
-    readarray -d '' FIELDS < <(git diff -z --diff-filter=AMRC --name-status "$@")
-    RESULT=()
-    # Simple "parser"
-    # The format is
-    # A|M <filename>
-    # R|C <old_filename> <new_filename>
-    set -- "${FIELDS[@]}"
-    while [[ $# -gt 0 ]]; do
-        type="$1"
-        case $type in
-            A|M)
-                RESULT+=("$2")
-                shift; shift;
-                ;;
-            R|C)
-                RESULT+=("$3")
-                shift; shift; shift;
-                ;;
-            *)
-                exit 1
-                ;;
-        esac
-    done
-    printf '%s\n' "${RESULT[@]}"
-}
 
-# git ls-files wrapper
-# wrap git ls-files to exclude deleted files
-function git_ls_files_wrapper() {
-    comm -23 <(git ls-files "$@" | sort) <(git ls-files --deleted | sort)
-}
+def run_parallel(func, objs_list, func_params=()):
```

Comment:
* So this runs stuff in parallel but only tasks of the same type, correct? When this kind of stuff happens I feel like we're reinventing `ninja`. But this is likely not the case. At very least, we can provide better reporting.
* Can we take this occasion to experiment with thread-less asyncio?
  There's an interesting [SO answer](https://stackoverflow.com/questions/48483348/how-to-limit-concurrency-with-python-asyncio) that offers many alternative solutions.
  This has been a recurring topic and I really think we should learn The Right Way™ to do non-CPU intensive coroutines in Python.

Reply:
I've reworked the passes a bit, now they're all `async` so a lot of threading code has been dropped
The only bit remaining that uses `run_parallel` is the one for reading the files, which does not have a `async` counterpart and *does* scale with python threads, so I've inlined the function and made it a bit less generic

Reply:
Nice. Out of curiosity, did measure that the reading the files was a bottleneck? Interesting.

Reply:
I've run a quick check, apparently `run_in_executor` has a bit of overhead, so when reading a lot of small files it's slower than reading them sequentially, whereas it's faster in the case of reading a few big files